src/client/runner/runner.js

```diff
@@ -80,44 +82,82 @@ Runner.prototype._onActionRun = function () {
     transport.asyncServiceMsg(msg);
 };
 
-Runner.prototype._onError = function (err) {
-    var runner = this;
+Runner.prototype._beforeScreenshot = function () {
+    this.stepIterator.suspend();
+    this.eventEmitter.emit(RunnerBase.SCREENSHOT_CREATING_STARTED_EVENT, {});
+    this.savedDocumentTitle = document.title;
 
-    if (this.stopped)
-        return;
+    var assignedTitle = `[ ${window.location.toString()} ]`;
 
-    //NOTE: we should stop stepIterator to prevent playback after an error is occurred
-    this.stepIterator.stop();
+    // NOTE: we should keep the page url in the
+    // document.title while screenshot creating run
+    this.checkTitleIntervalId = window.setInterval(() => {
+        if (document.title !== assignedTitle) {
+            this.savedDocumentTitle = document.title;
+            document.title          = assignedTitle;
+        }
+    }, 50);
 
-    RunnerBase.prototype._onError.call(this, err);
+    document.title = assignedTitle;
+
+    return new Promise(resolve => window.setTimeout(resolve), APPLY_DOCUMENT_TITLE_TIMEOUT);
+};
+
+Runner.prototype._afterScreenshot = function () {
+    window.clearInterval(this.checkTitleIntervalId);
+
+    document.title            = this.savedDocumentTitle;
+    this.checkTitleIntervalId = null;
+    this.savedDocumentTitle   = null;
+
+    this.eventEmitter.emit(RunnerBase.SCREENSHOT_CREATING_FINISHED_EVENT, {});
+    this.stepIterator.resume();
+
+    return new Promise(resolve => window.setTimeout(resolve), RESTORE_DOCUMENT_TITLE_TIMEOUT);
+};
 
-    if (!SETTINGS.get().TAKE_SCREENSHOT_ON_FAILS) {
-        this.stopped = true;
-        transport.fail(err, Runner.checkStatus);
+Runner.prototype._onTestFail = function (err, isAssertion) {
+    // NOTE: we should not create several screenshots for a step. Create a screenshot
+    // if it's the first error in the step or it's a error on page initialization.
+    err.pageUrl            = document.location.toString();
+    err.screenshotRequired = SETTINGS.get().TAKE_SCREENSHOTS && SETTINGS.get().TAKE_SCREENSHOT_ON_FAILS &&
+                             this.stepIterator.state.curStepErrors.length < 2;
+
+    return new Promise(resolve => {
+        if (err.screenshotRequired)
+            return this._beforeScreenshot().then(resolve);
+
+        resolve();
+    })
+        .then(() => new Promise(resolve => {
+            if (isAssertion)
+                transport.assertionFailed(err, resolve);
+            else
+                transport.fail(err, resolve);
```

Comment:
Whole this promise can be refactored to function, let's name it `_reportErrorToServer` or smthg